### PR TITLE
ImpervaWAF: dataconnector, parser, samples

### PR DESCRIPTION
**Pull request by averbn**
## Proposed Changes

  - add a new connector

### CODE REVIEW

**Comment by v-jayakal** [comment]
@oshezaf , can you please review the Parser, let us know. Thanks.

**Comment by v-admahe** [comment]
@averbn Could you please provide a sample data in json formate also include that file in PR.

Thanks,
Aditi

**Comment by averbn** [comment]
> @averbn Could you please provide a sample data in json formate also include that file in PR.
> 
> Thanks,
> Aditi

Data sample has been presented already in JSON format in this PR - .script/tests/KqlvalidationsTests/CustomTables/ImpervaWAFCloud_CL.json

**Comment by v-admahe** [comment]
> > @averbn Could you please provide a sample data in json formate also include that file in PR.
> > Thanks,
> > Aditi
> 
> Data sample has been presented already in JSON format in this PR - .script/tests/KqlvalidationsTests/CustomTables/ImpervaWAFCloud_CL.json

In Kqlvalidation we gave the structure of the table and in sample data folder we gave the sample data file.

Thanks,
Aditi

**Comment by v-admahe** [comment]
@averbn Package (Zip file) didn't has all the required file. Please check the sample file structure below.
[AzureFunctionProofpointTAP.zip](https://github.com/Azure/Azure-Sentinel/files/6764722/AzureFunctionProofpointTAP.zip)

Thanks,
Aditi

File: Solutions/ImpervaCloudWAF/Data Connectors/ImpervaWAFCloud_FunctionApp.json [hunk 1]
```diff
@@ -0,0 +1,121 @@
+{
+	"id": "ImpervaWAFCloudAPI",
+	"title": "Imperva Cloud WAF",
+	"publisher": "Atlassian",
+	"descriptionMarkdown": "The [Imperva Cloud WAF](https://www.imperva.com/resources/resource-library/datasheets/imperva-cloud-waf/) data connector provides the capability integrate and to ingest Web Aplication Firewall events into Azure Sentinel through the REST API. Refer to Log integration [documentation](https://docs.imperva.com/bundle/cloud-application-security/page/settings/log-integration.htm#Download) for more information. The connector provides ability to get events which helps to examine potential security risks, analyze your team's use of collaboration, diagnose configuration problems and more.",
+	"additionalRequirementBanner": "These queries and workbooks are dependent on a parser based on Kusto to work as expected. Follow the steps to use this Kusto functions alias **ImpervaWAFCloud** in queries and workbooks [Follow steps to get this Kusto functions>](https://aka.ms/sentinel-impervawafcloud-parser).",
```
Comment (v-maudan) [inline Solutions/ImpervaCloudWAF/Data Connectors/ImpervaWAFCloud_FunctionApp.json:6]: Please update additionalRequirementBanner as follows.
`">This data connector depends on a parser based on a Kusto Function to work as expected [**ImpervaWAFCloud**](https://aka.ms/sentinel-impervawafcloud-parser) which is deployed with the Azure Sentinel Solution."`

File: Solutions/ImpervaCloudWAF/Data Connectors/ImpervaWAFCloud_FunctionApp.json [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+	"instructionSteps": [{
+			"title": "",
+			"description": ">**NOTE:** This connector uses Azure Functions to connect to the Imperva Cloud API to pull its logs into Azure Sentinel. This might result in additional data ingestion costs. Check the [Azure Functions pricing page](https://azure.microsoft.com/pricing/details/functions/) for details."
+		},
+		{
+			"title": "",
+			"description": ">**(Optional Step)** Securely store workspace and API authorization key(s) or token(s) in Azure Key Vault. Azure Key Vault provides a secure mechanism to store and retrieve key values. [Follow these instructions](https://docs.microsoft.com/azure/app-service/app-service-key-vault-references) to use Azure Key Vault with an Azure Function App."
+		},
+		{
+			"description": ">**NOTE:** This data connector depends on a parser based on a Kusto Function to work as expected. [Follow these steps](https://aka.ms/sentinel-impervawafcloud-parser) to create the Kusto functions alias, **ImpervaWAFCloud**"
```
Comment (v-maudan) [inline Solutions/ImpervaCloudWAF/Data Connectors/ImpervaWAFCloud_FunctionApp.json:75]: Please update description as follows.
 `">**NOTE:**This data connector depends on a parser based on a Kusto Function to work as expected [**ImpervaWAFCloud**](https://aka.ms/sentinel-impervawafcloud-parser) which is deployed with the Azure Sentinel Solution."`

File: Solutions/ImpervaCloudWAF/Data Connectors/ImpervaWAFCloud_FunctionApp.json [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
...s** area, choose the **Deploy to function app** button.\nIf you aren't already signed in, choose the Azure icon in the Activity bar, then in the **Azure: Functions** area, choose **Sign in to Azure**\nIf you're already signed in, go to the next step.\n5. Provide the following information at the prompts:\n\n\ta. **Select folder:** Choose a folder from your workspace or browse to one that contains your function app.\n\n\tb. **Select Subscription:** Choose the subscription to use.\n\n\tc. Select **Create new Function App in Azure** (Don't choose the Advanced option)\n\n\td. **Enter a globally unique name for the function app:** Type a name that is valid in a URL path. The name you type is validated to make sure that it's unique in Azure Functions. (e.g. ImpervaCloudXXXXX).\n\n\te. **Select a runtime:** Choose Python 3.8.\n\n\tf. Select a location for new resources. For better performance and lower costs choose the same [region](https://azure.microsoft.com/regions/) where Azure Sentinel is located.\n\n6. Deployment will begin. A notification is displayed after your function app is created and the deployment package is applied.\n7. Go to Azure Portal for the Function App configuration."
```
Comment (v-maudan) [inline Solutions/ImpervaCloudWAF/Data Connectors/ImpervaWAFCloud_FunctionApp.json:114]: Please update Azure Function App Short link as follows:
https://aka.ms/sentinel-impervawafcloud-functionapp

File: Solutions/ImpervaCloudWAF/Data Connectors/azuredeploy_Connector_ImpervaWAFCloud_AzureFunction.json [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+                        "APPLICATIONINSIGHTS_CONNECTION_STRING": "[reference(resourceId('microsoft.insights/components', variables('FunctionName')), '2015-05-01').ConnectionString]",
+                        "AzureWebJobsStorage": "[concat('DefaultEndpointsProtocol=https;AccountName=', toLower(variables('FunctionName')),';AccountKey=',listKeys(resourceId('Microsoft.Storage/storageAccounts', toLower(variables('FunctionName'))), '2019-06-01').keys[0].value, ';EndpointSuffix=',toLower(variables('StorageSuffix')))]",
+                        "WorkspaceID": "[parameters('WorkspaceID')]",
+                        "WorkspaceKey": "[parameters('WorkspaceKey')]",
+                        "ImpervaAPIID": "[parameters('ImpervaAPIID')]",
+                        "ImpervaAPIKey": "[parameters('ImpervaAPIKey')]",
+                        "ImpervaLogServerURI": "[parameters('ImpervaLogServerURI')]",
+                        "logAnalyticsUri": "[variables('LogAnaltyicsUri')]",
+                        "WEBSITE_RUN_FROM_PACKAGE": "https://github.com/averbn/azure_sentinel_data_connectors/raw/main/imperva-waf-cloud-azure-sentinel-data-connector/ImpervaWAFCloudSentinelConn.zip"
```
Comment (v-maudan) [inline Solutions/ImpervaCloudWAF/Data Connectors/azuredeploy_Connector_ImpervaWAFCloud_AzureFunction.json:162]: Replace "WEBSITE_RUN_FROM_PACKAGE" value with https://aka.ms/sentinel-impervawafcloud-functionapp

**Review by v-maudan — CHANGES_REQUESTED** [review]

File: Solutions/ImpervaCloudWAF/Data Connectors/ImpervaWAFCloud_FunctionApp.json [hunk 5]
```diff
@@ -0,0 +1,121 @@
+{
+	"id": "ImpervaWAFCloudAPI",
+	"title": "Imperva Cloud WAF",
+	"publisher": "Atlassian",
+	"descriptionMarkdown": "The [Imperva Cloud WAF](https://www.imperva.com/resources/resource-library/datasheets/imperva-cloud-waf/) data connector provides the capability integrate and to ingest Web Aplication Firewall events into Azure Sentinel through the REST API. Refer to Log integration [documentation](https://docs.imperva.com/bundle/cloud-application-security/page/settings/log-integration.htm#Download) for more information. The connector provides ability to get events which helps to examine potential security risks, analyze your team's use of collaboration, diagnose configuration problems and more.",
```
[the remark this answers is not in the document]
Comment (averbn) [reply Solutions/ImpervaCloudWAF/Data Connectors/ImpervaWAFCloud_FunctionApp.json]: done

File: Solutions/ImpervaCloudWAF/Data Connectors/ImpervaWAFCloud_FunctionApp.json [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+						],
+						"label": "Primary Key"
+					},
+					"type": "CopyableLabel"
+				}
+			]
+		},
+		{
+			"title": "Option 1 - Azure Resource Manager (ARM) Template",
+			"description": "Use this method for automated deployment of the Imperva Cloud WAF data connector using an ARM Tempate.\n\n1. Click the **Deploy to Azure** button below. \n\n\t[![Deploy To Azure](https://aka.ms/deploytoazurebutton)](https://aka.ms/sentinelimpervacloudwafazuredeploy)\n2. Select the preferred **Subscription**, **Resource Group** and **Location**. \n> **NOTE:** Within the same resource group, you can't mix Windows and Linux apps in the same region. Select existing resource group without Windows apps in it or create new resource group.\n3. Enter the **ImpervaAPIID**, **ImpervaAPIKey**, **ImpervaLogServerURI** and deploy. \n4. Mark the checkbox labeled **I agree to the terms and conditions stated above**. \n5. Click **Purchase** to deploy."
```
[the remark this answers is not in the document]
Comment (averbn) [reply Solutions/ImpervaCloudWAF/Data Connectors/ImpervaWAFCloud_FunctionApp.json]: done

File: Solutions/ImpervaCloudWAF/Data Connectors/ImpervaWAFCloud_FunctionApp.json (hunk 1 above)
Comment (averbn) [reply Solutions/ImpervaCloudWAF/Data Connectors/ImpervaWAFCloud_FunctionApp.json]: done

File: Solutions/ImpervaCloudWAF/Data Connectors/ImpervaWAFCloud_FunctionApp.json (hunk 2 above)
Comment (averbn) [reply Solutions/ImpervaCloudWAF/Data Connectors/ImpervaWAFCloud_FunctionApp.json]: done

File: Solutions/ImpervaCloudWAF/Data Connectors/ImpervaWAFCloud_FunctionApp.json (hunk 3 above)
Comment (averbn) [reply Solutions/ImpervaCloudWAF/Data Connectors/ImpervaWAFCloud_FunctionApp.json]: done

File: Solutions/ImpervaCloudWAF/Data Connectors/azuredeploy_Connector_ImpervaWAFCloud_AzureFunction.json (hunk 4 above)
Comment (averbn) [reply Solutions/ImpervaCloudWAF/Data Connectors/azuredeploy_Connector_ImpervaWAFCloud_AzureFunction.json]: done

**Comment by v-admahe** [comment]
@averbn Parser is not working, it giving error. Please make the require change in the parser.

image
image

Thanks,
Aditi

**Comment by averbn** [comment]
> @averbn Parser is not working, it giving error. Please make the require change in the parser.
> 
> image
> image
> 
> Thanks,
> Aditi

Do you have logs in ImpervaWAFCloud_CL table? So the parser works properly in my side. Could you please share more details?
image

**Comment by v-admahe** [comment]
> @averbn Parser is not working, it giving error. Please make the require change in the parser.
> 
> image
> image
> 
> Thanks,
> Aditi

@averbn There are some field like postbody, qstr and ver are missing in the sample data this is the reason parser is failing.

Thanks
Aditi

File: Solutions/ImpervaCloudWAF/Parsers/ImpervaWAFCloud [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+                    replace(@"\s","",cs8Label), cs8
+                 )
+| evaluate bag_unpack(packed)
+| project-away logs, msg, cs1, cs1Label, cs2, cs2Label, cs3, cs3Label, cs4,cs4Label, cs5,cs5Label, cs6, cs6Label, cs7, cs7Label, cs8, cs8Label, Message
+| extend    EventVendor = "Imperva",
+            EventProduct = product,
+            EventType = type,
+            EventSeverity = severity,
+            DvcAction = act,
+            NetworkApplicationProtocol = app,
```
Comment (v-rucdu) [inline Solutions/ImpervaCloudWAF/Parsers/ImpervaWAFCloud:34]: The parser fails for below reasons,
1. There is an inbuilt function with same name as 'app' which cause ambiguity. Please refer below screenshot for the error
image

2. The sample data does not have values for **qstr**, **postbody** and **ver** fields

File: Solutions/ImpervaCloudWAF/Parsers/ImpervaWAFCloud [hunk 8]
```diff
@@ -0,0 +1,66 @@
+// Usage Instruction : 
+// Paste below query in log analytics, click on Save button and select as Function from drop down by specifying function name and alias as ImpervaWAFCloud.
+// Function usually takes 10-15 minutes to activate. You can then use function alias from any other queries (e.g. JiraAudit | take 10).
```
Comment (v-rucdu) [inline Solutions/ImpervaCloudWAF/Parsers/ImpervaWAFCloud:3]: Please update function alias

File: Solutions/ImpervaCloudWAF/Data Connectors/ImpervaWAFCloud_FunctionApp.json [hunk 9]
```diff
@@ -0,0 +1,121 @@
+{
+	"id": "ImpervaWAFCloudAPI",
+	"title": "Imperva Cloud WAF",
+	"publisher": "Atlassian",
+	"descriptionMarkdown": "The [Imperva Cloud WAF](https://www.imperva.com/resources/resource-library/datasheets/imperva-cloud-waf/) data connector provides the capability integrate and to ingest Web Application Firewall events into Azure Sentinel through the REST API. Refer to Log integration [documentation](https://docs.imperva.com/bundle/cloud-application-security/page/settings/log-integration.htm#Download) for more information. The connector provides ability to get events which helps to examine potential security risks, analyze your team's use of collaboration, diagnose configuration problems and more.",
```
Comment (v-rucdu) [inline Solutions/ImpervaCloudWAF/Data Connectors/ImpervaWAFCloud_FunctionApp.json:5]: Shouldn't the description text start be like ,

The Imperva Cloud WAF data connector provides the capability **to** integrate and ~~to~~ ingest Web Application Firewall events into Azure Sentinel through the REST API.

**Review by v-rucdu — CHANGES_REQUESTED** [review]

**Comment by v-rucdu** [comment]
> @oshezaf , can you please review the Parser, let us know. Thanks.

@oshezaf : Could you please review parser... Thanks!!!

File: Solutions/ImpervaCloudWAF/Parsers/ImpervaWAFCloud (hunk 7 above)
Comment (averbn) [reply Solutions/ImpervaCloudWAF/Parsers/ImpervaWAFCloud:34]: 1- The "app" field name is automatically parsed from predefined CEF log format, I can`t change this name. It is underlined but works as expected. 
2- I Added the samples with qstr, postbody and ver fields. Pls check again

File: Solutions/ImpervaCloudWAF/Data Connectors/ImpervaWAFCloud_FunctionApp.json (hunk 9 above)
Comment (averbn) [reply Solutions/ImpervaCloudWAF/Data Connectors/ImpervaWAFCloud_FunctionApp.json:5]: done

File: Solutions/ImpervaCloudWAF/Parsers/ImpervaWAFCloud (hunk 8 above)
Comment (averbn) [reply Solutions/ImpervaCloudWAF/Parsers/ImpervaWAFCloud:3]: done

File: Solutions/ImpervaCloudWAF/Parsers/ImpervaWAFCloud (hunk 7 above)
Comment (v-rucdu) [reply Solutions/ImpervaCloudWAF/Parsers/ImpervaWAFCloud:34]: Thanks @averbn, the parser is working fine now

**Comment by v-rucdu** [comment]
Signing-off for DataConnector

**Review by v-jayakal — APPROVED** [review]

**Pull request merged**